### ISSUE TITLE
No response after selecting text in a Chinese book

### ISSUE THREAD

**Issue by tete1030** [koreader/koreader#9581]
* KOReader version: 2022.08
* Device: Kindle Oasis 3

#### Issue

Once I long press on any page, the whole koreader and system freeze. Pressing power button only switches frontlight. 

#### Steps to reproduce

1. Open the book 
2. long press on texts

##### `crash.log` (if applicable)

no crash.log

Book download: 
https://mega.nz/file/5URVwZJL#5WhPmUflF1-0JA-Dn5tHV1M6GEjk288uiKn4a8Ysyk8

**Comment by tete1030** [comment]
Video:

https://user-images.githubusercontent.com/4312115/193452025-b8a228d4-6f15-4f29-ac87-002286699217.MOV

**Comment by tete1030** [comment]
Update: After posting this issue, the problem has gone magically (only a few seconds delay on long pressing and selecting). But in previous cases I have waited longer than 10 minutes with no response, I had to reboot it.

**Comment by poire-z** [comment]
No issue for me with this book, long-pressing on a char and launching a dict lookup.
Sometimes, a delay can be caused by crappy HTML - but it feels it is not the case with this book (althought SumatraPDF has a hard time opening this EPUB).

Did it happen with this book only ? or any book.
Next time you get the slow down, see options in `Tap and gestures > Long-press on text >`, like unchecking `[ ] Dict on single word selection` - to guess where the slowdown is happening (text slection drawing, dictionary lookup....)

**Comment by NiLuJe** [comment]
Also, as usual, verbose debug logs might come in handy ;).

([Tools] > More tools > Developer options).

**Comment by tete1030** [comment]
[crash.log](https://github.com/koreader/koreader/files/9694547/crash.log)

The stuck begins at 10/03/22-11:52:59, last for 3.5 minutes

**Comment by tete1030** [comment]
> Did it happen with this book only ? or any book. Next time you get the slow down, see options in `Tap and gestures > Long-press on text >`, like unchecking `[ ] Dict on single word selection` - to guess where the slowdown is happening (text slection drawing, dictionary lookup....)

Tried another Chinese book with `Japanese support` turned on, still the problem. Sorry I cannot find any `Long-press on text` or `Dict on single word selection` in the `Tap and gestures` menu

**Comment by poire-z** [comment]
Thanks. Any reason you had Japanese support enabled ? Do you read books in both languages?

Pinging @cyphar (the author of Japanese support): can you reproduce ? Is there some issue when using it on non-Japanese which should be taken care of ?

**Comment by tete1030** [comment]
No I don't read Japanese. Guess I was kind of enthusiastic about 'enabling all the features' in case one day I can read Japanese :D, turned on all plugins available in the Plugin Management.

**Comment by tete1030** [comment]
Located the culprit.

The Japanese plugin tries to call `sdcv`, likely enumerating the dict data dir. I have a huge `res` dir inside this dir, containing 7000+ files. This enumerating process seems to be very very slow even with `ls` command. Moving out `res` solves the problem. 

In some cases the stuck was not longer than a few seconds. This could be explained by filesystem cache.

Closing this issue. Thanks for your help

**Comment by cyphar** [comment]
Yeah, so the thing is the Japanese support plugin is most likely also useful for Chinese readers (it will auto-select the longest word available in any dictionary you have installed, and the need for this with Chinese is similar to the need for this in Japanese though Japanese requires more work because of verb inflections).

But yeah if sdcv struggles with your dict directory that's more of an issue with sdcv.

**Comment by NiLuJe** [comment]
> I have a huge `res` dir inside this dir, containing 7000+ files.

FWIW, Kindles are very very very bad at this. First of all, because FAT32 (but that's true of most of our platforms), but specifically because of the userstore fuse proxy ;).

(IIRC, the stock OS starts shouting at you if a directory has more than a thousand entries or something).

**Comment by tete1030** [comment]
> Yeah, so the thing is the Japanese support plugin is most likely also useful for Chinese readers (it will auto-select the longest word available in any dictionary you have installed, and the need for this with Chinese is similar to the need for this in Japanese though Japanese requires more work because of verb inflections).

I think you are right. It should work the same way for both languages.

I did disable the dictionary referencing the `res` dir, but this setting was not respected. Besides, it's better if the plugin could distinguish the language of dicts according to the book's language, rather than indexing through all dicts. Is this possible?

**Comment by tete1030** [comment]
> > I have a huge `res` dir inside this dir, containing 7000+ files.
> 
> FWIW, Kindles are very very very bad at this. First of all, because FAT32 (but that's true of most of our platforms), but specifically because of the userstore fuse proxy ;).
> 
> (IIRC, the stock OS starts shouting at you if a directory has more than a thousand entries or something).

Good to know, thanks. The access speed is much faster under USB storage mode from my computer. I guess the fuse proxy was not in effect under this mode? or just cache again.

**Comment by NiLuJe** [comment]
Yeah, you get access to the raw FAT32 fs over USBMS ;).

**Comment by mergen3107** [comment]
@NiLuJe 
I can see from here (https://kb.taktpraha.cz/projects/kindlepw5) that PW5 has eMMC 5.1. Is this decent already, or still slow? I mean, ofc compared to UFS 2.0 it is slower especially in Random IOPS, but maybe there is improvement over PW3/4?

**Comment by poire-z** [comment]
> I did disable the dictionary referencing the res dir, but this setting was not respected.

That just tells which dictionary we should tell sdcv to _not use_.
sdcv, and koreader (to build the list of available dicts), both scan all subdirectories (because people could put dicts inside dicts - we don't prevent that, sdcv does not mind). So, we enter and scan res/ for any .ifo or any subdirectory.
We didn't need to bother until now :) but well, users are clever: you noticed you had a huge res/ - and I give you a new info: sdcv & koreader don't use what's in it, so you can delete its content :)

> Besides, it's better if the plugin could distinguish the language of dicts according to the book's language, rather than indexing through all dicts. Is this possible?

Stardicts dicts don't advertize their language(s) in the .ifo. So, we don't know. Also, books may have sentences in multiple languages, and having and using multiple language dicts will have the good language dicts knowing about that word give you some results.

**Comment by NiLuJe** [comment]
> @NiLuJe I can see from here (https://kb.taktpraha.cz/projects/kindlepw5) that PW5 has eMMC 5.1. Is this decent already, or still slow? I mean, ofc compared to UFS 2.0 it is slower especially in Random IOPS, but maybe there is improvement over PW3/4?

I wasn't necessarily impressed, but then I didn't really test much (and the last Kindle I had before it was a PW2 ;p).

**Comment by mergen3107** [comment]
@NiLuJe 
Ok, thanks :)
